README.md

# NeoServer

**NeoServer** is a Node.js module to create a web server easily.

## How to install

To install NeoServer, you first needs to install the [NodeJS Package Manager (npm)](https://www.npmjs.com/).

Then, type the following command in your terminal :

```sh
npm install neo-server -g
```

**NOTE :** The *-g* argument install NeoServer globally. This permit to use NeoServer everywhere and access to the 'neo' command.

## Create your server

Now, let's create your own server !
First, create an empty folder anywhere on your computer, and then go into using a terminal.

```sh
mkdir neo-site
cd neo-site
```

Create a first page, named **index.html**, and put into a content like *Hello world !*

Then, run the server.

```sh
neo
```

The server will be launched on port **8080**. Open a browser like Firefox, and type in the URL bar :
http://localhost:8080
And see the result !

## Command-line options

Argument                | Default    | Description
----------------------- | ---------- | -----------
--port                  | 8080       | Port to run the server
--encoding              | utf-8      | The web pages encoding
--rootfile              | index.html | Files to use when the URL terminates by a slash
--disable-neoserv-files | false      | Set to "true" to disable *.neoserv* files execution
--log-requests          |            | Set to a value to log all incoming requests and store it in the specified file
--dynamic-neoserv-file  | false      | Set to "true" to reload *.neoserv* file at each incoming requests
--benchmark             | false      | Set to "true" to display time spent to send the request (in miliseconds)
--debug                 |            | Include this argument to see errors' details
--mods                  |            | List of mods, all names separated by a comma

## *.neoserv* files

*.neoserv* files permit to perform actions on the server, like *.htaccess* on **Apache**.
A *.neoserv* file is organized around *categories* and *parameters*. There is one thing by line.

Example :

```neoserv
rootfile = index.html,index.htm
xcutable = html,htm

[mime-types]
jsp = application/javascript

[deny]
private/

[errorDocument]
ErrorDocument 404 errors/404.html

[rewrite]
RewriteUrl errors/* errors/$1.html
```

Let's see how it does...
First, we define two **globals** parameters : *rootfile* and *xcutable*.

*rootfile* permit to override the same command-line parameter. File names all separated by commas.
*xcutable* define files which can have inline **runable** JavaScript content.

Next, this file define a *category* : *mime-types*.
This category permit to define the [Mime-Type](https://en.wikipedia.org/wiki/MIME) of different file extensions.
Here, there is a parameter to define *.jsp* extension as *application/javascript* mime-type.

Next, the *deny category*, that make client unable to access to certain URLs.
To forbid access to an entire directory, make the name terminate by a slash, like here.

The *errorDocument* indicate what HTML page use when there is an error. If you don't specify an error document, the default document is used : a big title that indicate there was an error.
Here, the page *errors/404.html* is called each time there is a **404** (Not Found) error.

Finally, the *rewrite category*, the most powerful too, permit to redirect URLs to others. The URL in the client's URL bar will not be changed, but the server will deliver the content of the specified page.
For example, if the client access to *errors/500*, the server will deliver the content of the page *errors/500.html*.

## Use inline JavaScript

In an executable web page (default : all files with *.html* extension), you can use **inline JavaScript**.
That's a JS code which perform operations. Example of using :

```html
<p>You are seeing the URL : <% request.url %></p>
```

The code between *<%* and *%>* is runned and the result is placed in the page. For example, if the URL is **documents/interview.html**, the displayed content will be :

```html
<p>You are seeing the URL : documents/interview.html</p>
```

You can access to only three variables : *require*, *request* and *response*. *request* and *response* are Node.js variables.


CHANGELOG.md
# NeoServer changelog

## NeoServer v1.1
- New : Support of "?" and "#" in URLs
- New : Added mods feature
- New : Allow **return** statement in inline-JS
- New : Added *echo* (alias of *print*) command for inline-JS
[Bug fixes]
- Fixed bug M003 : Multiple lines JS were not detected
- Fixed bug M002 : Server crashed if you don't specify *xcutable* parameter in your *.neoserv* file
- Fixed bug M001 : URL Rewriting was not working when the final URL not starting by a slash
[Notes]
- *skyer.js* example doesn't work for the moment

## NeoServer v1.0
- New : Serve web pages
- New : Serve binary files, including Content-Length
- New : Support of *.neoserv* files, including URL-Rewriting
- New : Support of command-line configuration
- New : Can log each request
- New : Command-line benchmark
- New : Support of Error Documents
- New : Support of inline JS in executables web pages


mods/php_js/mod.js

exports.require('die');
exports.require('include');

exports.runtime = function(context) {
    var php = require('phpjs'), fs = require('fs'), cmv = context.vm_context;

    function normalize(file) {
        var o_file = file;
        file = path.normalize(file);
        var resolved = path.resolve(file), cwd = process.cwd();

        if(resolved.substr(0, cwd.length) !== cwd) {
            var parts = file.split('/'),
                safe  = [];

            for(var i = 0; i < parts.length; i += 1) {
                parts[i] = parts[i].replace(/[^a-zA-Z0-9_\-\+\/ "'\:\,\.\;\{\}\(\)\[\]]/g, '');
                if (!parts[i] || ('.' == parts[i])) {
                    continue;
                } else if('..' == parts[i]) {
                    safe.pop();
                    continue;
                } else {
                    safe.push(parts[i]);
                }
            }

            // Return the "clean" file
            file     = path.normalize(safe.join('/'));
            resolved = path.resolve(file);

            if(resolved.substr(0, cwd.length) !== cwd)
                throw new Error(colors.red('Unable to make path secured : ' + colors.cyan(o_file) + (argument('d', 'debug') ? '\nResolved : ' + resolved + '\nFinal : ' + file : '')));
        }

        return path.join(process.cwd(), (file.substr(0, 1) === '/' ? file.substr(1) : file));
    }

    /*for(var i in cmv)
        php[i] = cmv[i];

    cmv = php;*/

    for(var i in php)
        cmv[i] = php[i];

    cmv.is_dir = function(dir) {
        return cmv.file_exists(dir) && !fs.lstatSync(dir).isFile()
    };

    cmv.is_file = function(file) {
        return cmv.file_exists(file) && fs.lstatSync(file).isFile()
    };

    cmv.file_exists = function(file) {
        return fs.existsSync(normalize(file));
    };

    cmv.file_get_contents = function(file, charset) {
        if(!cmv.is_file(file))
            throw new Error('File not found ' + file);

        fs.readFileSync(normalize(file), charset || 'utf-8'); return ;
    };

    cmv.file_put_contents = function(file, content, charset) {
        fs.writeFileSync(normalize(file), content, charset || 'utf-8'); return ;
    };

    cmv.mkdir = function(dir) {
        fs.mkdirSync(normalize(dir)); return ;
    };

    cmv.unlink = function(file) {
        if(!cmv.is_file(file))
            throw new Error('File not found ' + file);

        fs.unlinkSync(file);
    };

};


neo.js
#!/usr/bin/env node

function argument(_short, _long) {
    return (_short ? short[_short] : undefined) || (_long ? long[_long] : undefined);
}

var unsorted_args = process.argv.splice(2);

var args  = [];
var short = {};
var long  = {};

for(var i = 0; i < unsorted_args.length; i++) {
	if(unsorted_args[i].substr(0, 1) === '-') {
		var isLong = unsorted_args[i].substr(0, 2) === '--';
		var hasEqual = unsorted_args[i].indexOf('=') !== -1;
		(isLong ? long : short)[hasEqual ? unsorted_args[i].substr(1 + isLong, unsorted_args[i].indexOf('=') - 1 - isLong) : unsorted_args[i].substr(1 + isLong)] = (hasEqual ? unsorted_args[i].substr(unsorted_args[i].indexOf('=') + 1) : true);
	} else
		args.push(unsorted_args[i]);
}

var http         = require('http'),
    fs           = require('fs'),
    path         = require('path'),
    colors       = require('chalk'),
    vm           = require('vm'),
    isbinaryfile = require('isbinaryfile'),
    cmdArgs      = process.argv.splice(2), config = {
        port     : argument('p', 'port') || 8080,
        encoding : argument('e', 'encoding') || 'utf-8',
        rootfile : (argument('i', 'index') || 'index.html').split(',')
    }, neoServ, mods = {}, events = {};

function load_neoserv_file() {
    neoServ     = {
        'errors-code': {
            403: 'Forbidden',
            404: 'Not Found',
            500: 'Internal Server Error'
        },

        'mime-types': {
            html: 'text/html',
            htm : 'text/html',
            css : 'text/css',
            xml : 'text/xml',
            txt : 'text/plain',
            csv : 'text/csv',

            js  : 'application/javascript',
            zip : 'application/zip',
            json: 'application/json',
            pdf : 'application/pdf',
            ogg : 'application/ogg',
            sfw : 'application/x-shockwave-flash',

            gif : 'image/gif',
            jpg : 'image/jpeg',
            jpeg: 'image/jpeg',
            png : 'image/png',
            tiff: 'image/tiff',
            svg : 'image/svg+xml',

            mp3 : 'audio/mpeg',
            wav : 'audio/wav',
            wma : 'audio/x-ms-wma',

            mp4 : 'video/mp4',
            wmv : 'video/x-ms-wmv',
            avi : 'video/x-msvideo',
            flv : 'video/x-flv',
            webm: 'video/webm',

            xls : 'application/vnd.ms-excel',
            xlsx: 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet',
            ppt : 'application/vnd.ms-powerpoint',
            pptx: 'application/vnd.openxmlformats-officedocument.presentationml.presentation',
            doc : 'application/msword',
            docx: 'application/vnd.openxmlformats-officedocument.wordprocessingml.document',
            xul : 'application/vnd.mozilla.xul+xml',

            odt : 'application/vnd.oasis.opendocument.text',
            ods : 'application/vnd.oasis.opendocument.spreadsheet',
            odp : 'application/vnd.oasis.opendocument.presentation',
            odg : 'application/vnd.oasis.opendocument.graphics',
            odc : 'application/vnd.oasis.opendocument.chart',
            odf : 'application/vnd.oasis.opendocument.formula',
            odb : 'application/vnd.oasis.opendocument.database',
            odi : 'application/vnd.oasis.opendocument.image',
            odm : 'application/vnd.oasis.opendocument.text-master',

            ott : 'application/vnd.oasis.opendocument.text-template',
            ots : 'application/vnd.oasis.opendocument.spreadsheet-template',
            otp : 'application/vnd.oasis.opendocument.spreadsheet-template',
            otg : 'application/vnd.oasis.opendocument.graphics-template',

            '.binary': 'application/octet-stream',
            'default': 'text/plain'
        },

        xcutable: ['html']
    };

    if(fs.existsSync('.neoserv') && fs.lstatSync('.neoserv').isFile() && argument('n', 'disable-neoserv-files') !== 'true') {
        try {
            var neofile = fs
                .readFileSync('.neoserv', 'utf-8')
                .replace(/###((.|\n)*?)###/g, '')
                .replace(/#(.*)$/mg, '')
                .split('\n'),
                line, value, category;

            for(var i = 0; i < neofile.length; i += 1) {
                line = neofile[i].trim();

                if(!line)
                    continue;

                if(line.substr(0, 1) === '[' && line.substr(line.length - 1, 1)) {
                    category = line.substr(1, line.length - 2);

                    if(!neoServ.hasOwnProperty(category) && typeof neoServ[category] !== 'undefined')
                        throw new Error(colors.red(colors.cyan('.neoserv') + ' : System-reserved category name at line ' + + colors.green(i + 1)));

                    if(!neoServ.hasOwnProperty(category))
                        neoServ[category] = {};
                } else {
                    line  = line.replace(/^([a-zA-Z_]+)( +)([^ =]+)( +)([^ ]+)$/, '$3=$5');

                    if(line.indexOf('=') === -1) {
                        if(category) {
                            if(!Array.isArray(neoServ[category])) {
                                if(Object.keys(neoServ[category]).length)
                                    throw new Error(colors.red(colors.cyan('.neoserv') + ' : Can\'t use un-assignemnt in assignment category at line ' + colors.green(i + 1)));

                                neoServ[category] = [];
                            }

                            neoServ[category].push(line);
                        } else
                            throw new Error(colors.red(colors.cyan('.neoserv') + ' : Bad syntax at line ' + colors.green(i + 1)));
                    } else {
                        value = line.substr(line.indexOf('=') + 1).trim();
                        line  = line.substr(0, line.indexOf('=')).trim();

                        if(!category) {
                            neoServ[line] = value;
                        } else {
                            if(Array.isArray(neoServ[category]))
                                throw new Error(colors.red(colors.cyan('.neoserv') + ' : Can\'t use assignemnt in un-assignment category at line ' + colors.green(i + 1)));

                            neoServ[category][line] = value;
                        }
                    }
                }
            }

            if(neoServ.hasOwnProperty('rootfile') && neoServ.replaceConfig !== 'false')
                config.rootfile = neoServ.rootfile.split(',');

            if(neoServ.hasOwnProperty('xcutable') && typeof neoServ.xcutable === 'string')
                config.xcutable = neoServ.xcutable.split(',');

            if(neoServ.hasOwnProperty('rewrite') && !Array.isArray(neoServ.rewrite)) {
                var frewrite = [];

                for(var i in neoServ.rewrite) {
                    frewrite.push([
                        new RegExp('^\\/' + i
                            .replace(/([.*+?^=!:${}()|\[\]\/\\])/g, "\\$1")
                            .replace(/\\\?/g, '(.)')
                            .replace(/\\\*/g, '(.*?)') + '$'),
                        neoServ.rewrite[i]
                    ]);
                }

                neoServ.rewrite = frewrite;
            }

            if(neoServ.hasOwnProperty('deny') && Array.isArray(neoServ.deny)) {
                for(var i = 0; i < neoServ.deny.length; i += 1) {
                    neoServ.deny[i] = path.normalize(neoServ.deny[i]);
                }
            }
        }

        catch(e) {
            console.error(colors.red('Failed to load ' + colors.cyan('.neoserv') + ' file'), argument('d', 'debug') ? e.stack : e);
        }
    }
}

var load_mods;

if(load_mods = argument('m', 'mods')) {
    if(typeof load_mods !== 'string')
        throw new Error(colors.red(colors.cyan('mods') + ' argument must be a string'));

    var dir, mod, exports;
    load_mods = load_mods.split(',')

    for(var i = 0; i < load_mods.length; i += 1) {
        if(!mods.hasOwnProperty(load_mods[i])) {
            dir = path.join(__dirname, 'mods', load_mods[i]);

            try {
                if(!fs.existsSync(dir) || fs.lstatSync(dir).isFile()) {
                    console.error(colors.red('Can\'t load mod ' + colors.cyan(load_mods[i]) + ' : module not found'));
                    process.exit();
                }
            }

            catch(e) {
                console.error(colors.red('Can\'t load mod ' + colors.cyan(load_mods[i]) + ' : Can\'t access to ' + colors.green(__dirname) + (argument('d', 'debug') ? '\n' + e.stack : '')));
                process.exit();
            }

            try {
                mod = fs.readFileSync(path.join(dir, 'mod.js'), 'utf-8');
            }

            catch(e) {
                console.error(colors.red('Can\'t load mod ' + colors.cyan(load_mods[i]) + ' : Can\'t access to ' + colors.green(path.join(dir, 'mod.js') + (argument('d', 'debug') ? '\n' + e.stack : ''))));
                process.exit();
            }

            try {
                exports = mods[load_mods[i]] = {
                    on: function(name, callback) {
                        return typeof callback === 'function' ? events[name] = callback : events[name];
                    },

                    require: function(mod) {
                        load_mods.push(mod);
                    }
                };

                eval(mod);
            }

            catch(e) {
                console.error(colors.red('Can\'t load mod ' + colors.cyan(load_mods[i]) + ' : Runtime error' + (argument('d', 'debug') ? '\n' + e.stack : '')));
                process.exit();
            }
        }
    }
}

load_neoserv_file();
var httpRequest;

function call_event(name, args, additionnalArgs) {
    if(!events[name])
        return false;

    if(typeof additionnalArgs === 'object') {
        for(var i in additionnalArgs) {
            args[i] = additionnalArgs[i];
        }
    }

    return events[name](args);
};

function denied(file) {
    if(!neoServ.hasOwnProperty('deny') || !Array.isArray(neoServ.deny))
        return false;

    file = path.normalize(file);

    for(var i = 0; i < neoServ.deny.length; i += 1) {
        if(file.substr(0, neoServ.deny[i].length) === neoServ.deny[i])
            return true;
    }

    return false;
}

function neo_request(file, content, req, code) {
    var request = req.request, response = req.response, neoServ = req.neoServ,
        split   = file.split('/');

    if(denied(file) || split[split.length - 1] === '.neoserv') {
        error(403);
        return false;
    }

    var ext = split[split.length - 1];

    var s = ext.split('.');
    ext = s.length > 1 ? s[s.length - 1] : false;

    if(content instanceof Buffer) {
        response.writeHead(code || 200, {
            'Content-Type': neoServ['mime-types'][ext] || neoServ['mime-types']['.default'],
            'Content-Length': content.length
        });

        response.end(content);
        return true;
    } else {
        if(neoServ.xcutable.indexOf(ext) !== -1) {
            /*var ctx = vm.createContext({
                request: request,
                require: require
            });*/

            var cut = content.split('\n'), index, runtime = {
                stopExec: false
            };

            for(var i = 0; i < cut.length; i += 1) {
                /*if((index = cut[i].indexOf('<%')) !== -1)
                    cut[i] = cut[i].substr(0, index) + '<%' + (i + 1) + cut[i].substr(index + 2);*/
                cut[i] = cut[i].replace(/<%/g, '<%' + (i + 1));
            }

            content  = cut.join('\n');

            content  = content.replace(/<%([0-9]+)((.|\n)*?)%>/g, function(match, line, neo) {
                if(runtime.stopExec)
                    return '';

                try {
                    //var c = vm.runInContext(neo, ctx);
                    /*var c = (new vm.Script(neo, {
                        filename: file,
                        displayErrors: false
                    })).runInContext(ctx);*/

                    function get_mod_context() {
                        return {
                            request      : request,
                            response     : response,
                            line         : line,
                            vm_script    : script,
                            vm_code      : neo,
                            vm_context   : ctx,
                            runtime      : runtime
                        }
                    }

                    var ctx = {
                        request : request ,
                        response: response,
                        mods    : mods,
                        echo    : function(content) {
                            insert += content;
                        }
                    }, script = vm.createScript(neo.split('\n').length > 1 ? '(function() { /*function echo(content) { insert += content; } function print(content) { insert += content; }*/ ' + neo + '})();' : neo, file);

                    for(var i in mods) {
                        if(mods[i].runtime) {
                            exports = mods[i];
                            mods[i].runtime(get_mod_context(), mods[i]);
                        }
                    }

                    var insert = '', c = script.runInNewContext(ctx, file);
                    call_event('vm_end', get_mod_context(), {result: c, insert: insert, final_result: insert || c});

                    c     = insert || c; // doesn't work

                    return (c || typeof c === 'string' || typeof c === 'number') ? c : '';
                }

                catch(e) {
                    var o = e;
                    e = e.stack.split('\n');
                    var stack = e[0], el, lin, col, fil, already = false, backtrace = [];

                    for(var i = 1; i < e.length; i += 1) {
                        if(e[i].trim().substr(0, file.length + 4) !== 'at ' + file + ':' && already)
                            break;

                        fil = null;

                        if(e[i].trim().substr(0, file.length + 4) === 'at ' + file + ':') {
                            fil     = file;
                            already = true;
                        }

                        el  = e[i].trim().substr(3);
                        lin = parseInt(el.substr(el.indexOf(':') + 1, el.lastIndexOf(':') - el.indexOf(':') - 1)) + (fil ? line - 1 : 0);
                        col = el.substr(el.lastIndexOf(':') + 1);
                        fil = el.substr(0, el.indexOf(':'));

                        backtrace.push({
                            file: file,
                            line: lin,
                            col : col
                        });

                        stack += '<br />&nbsp;&nbsp;&nbsp;&nbsp;at ' + fil + ':' + lin + ':' + col;
                    }

                    var ret = call_event('vm_error', get_mod_context(), {
                        backtrace: backtrace,
                        stack: stack,
                        error: e,
                        message: o.message,
                        insert: insert
                    });

                    if(typeof ret !== 'undefined')
                        return ret;

                    if(!already) { // bug protection
                        stack = e[0] + '<br />&nbsp;&nbsp;&nbsp;&nbsp;<em>Unknwon stack</em>';
                        console.log(e.join('\n'));
                    }

                    return '<div class="neo-error" style="background-color:orange;border:1px solid black;padding:5px;padding-left:15px;font-family:monospace;">' + stack + '</div>';
                }
            });
        }

        if(response) {
            response.writeHead(code || 200, {
                'Content-Type': neoServ['mime-types'][ext] || neoServ['mime-types']['.default'],
                'Content-Length': content.length
            });

            call_event('send_response', {content: content});

            response.end(content);

            return true;
        } else {
            return content;
        }
    }
}

http.createServer(httpRequest = function(request, response) {

    var receivedRequest = Date.now();

    if(argument(false, 'log-requests')) {
        try {
            fs.appendFileSync(argument(false, 'log-requests'), Date.now() + ' ' + request.url + '\n', 'utf-8');
        }

        catch(e) {
            console.log(colors.red('Failed to store log :\n' + e.stack));
        }
    }

    if(argument('n', 'dynamic-neoserv-file') !== 'false')
        load_neoserv_file();

    function send(file, content, code) {
        return neo_request(file, content, {
            request: request,
            response: response,
            mods: mods,
            neoServ: neoServ
        }, code);
    }

    function error(errCode) {

        console.error(colors.red(errCode + ' ' + (neoServ['errors-code'][errCode] || '') + ' : ') + colors.green(request.treatUrl));

        if(neoServ.hasOwnProperty('errorDocument') && neoServ.errorDocument.hasOwnProperty(errCode)) {
            try {
                var content = fs.readFileSync(neoServ.errorDocument[errCode], neoServ.encoding || config.encoding);
            }

            catch(e) {
                console.error(colors.red('Cannot read ' + errCode + ' error template (') + colors.green(neoServ.errorDocument[errCode]) + colors.red(')' + (neoServ.verbose ? e.stack : '')));
            }

            if(typeof content !== 'undefined') {
                send(neoServ.errorDocument[errCode], content, errCode);
                console.info(colors.blue('Delivered ' + errCode + ' template : ') + colors.green('/' + neoServ.errorDocument[errCode]));
                return ;
            }
        }

        send(errCode + '.html', '<h1>' + errCode + ' ' + (neoServ['errors-code'][errCode] || '') + '</h1>', errCode);
        return ;

    }

    var final;
    request.treatUrl = decodeURI(request.url.replace(/(\?|#)(.*)$/, ''));

    if(neoServ.hasOwnProperty('rewrite') && Array.isArray(neoServ.rewrite)) {
        for(var i = 0; i < neoServ.rewrite.length; i += 1) {
            if((final = request.url.replace(neoServ.rewrite[i][0], neoServ.rewrite[i][1])) !== request.url) {
                //request.treatUrl = '/' + request.treatUrl.replace(neoServ.rewrite[i][0], neoServ.rewrite[i][1]);
                request.treatUrl = (final.substr(0, 1) === '/' ? '' : '/') + final;
                break;
            }
        }

        if(request.treatUrl !== request.url)
            call_event('url_rewrited', {origin: request.url, rewrited: request.treatUrl, request: request});
    }

    var dest = path.normalize(request.treatUrl).substr(1);

    try {
        var dest_exists = fs.existsSync(dest);
    }

    catch(e) {
        console.error(colors.red('Failed to check if requested file exists [' + colors.cyan(request.treatUrl) + '] [' + colors.green(dest) + ']') + (argument('d', 'debug') ? '\n' + e.stack : ''));
        error(500);

        if(argument('b', 'benchmark') !== 'false')
            console.log('Request sent in ' + (Date.now() - receivedRequest) + ' ms');

        return ;
    }

    if(!dest_exists) {
        error(404);

        if(argument('b', 'benchmark') !== 'false')
            console.log('Request sent in ' + (Date.now() - receivedRequest) + ' ms');

        return ;
    }

    try {
        var dest_is_directory = !fs.lstatSync(dest).isFile();
    }

    catch(e) {
        console.error(colors.red('Failed to check if requested resource is a directory [' + colors.cyan(request.treatUrl) + '] [' + colors.green(dest) + ']') + (argument('d', 'debug') ? '\n' + e.stack : ''));
        error(500);

        if(argument('b', 'benchmark') !== 'false')
            console.log('Request sent in ' + (Date.now() - receivedRequest) + ' ms');

        return ;
    }

    if(dest_is_directory) {
        var file, content;

        for(var i in config.rootfile) {
            try {
                file = path.join(dest, config.rootfile[i]);
                content = fs.readFileSync(file, neoServ.encoding || config.encoding);
            }

            catch(e) {
                if(cmdArgs.verbose)
                    console.error(colors.red('Failed to read index file : ') + colors.green(file));
            }

            if(typeof content !== 'undefined') {
                call_event('rootfile_used', {url: request.treatUrl, destination: dest, file: file, content: content, request: request})
                send(dest, content);
                console.info(colors.blue('Delivered : ') + colors.green('/') + ' [' + colors.cyan(file) + ']');
                return ;
            }
        }

        // list all files in the directory
        console.log('List all files...');

        error(404);

        if(argument('b', 'benchmark') !== 'false')
            console.log('Request sent in ' + (Date.now() - receivedRequest) + ' ms');

        call_event('request_sent_no_error', {url: request.treatUrl, destination: dest, duration: (Date.now() - receivedRequest), request: request});

        return ;
    }

    try {
        var content  = fs.readFileSync(dest);
        var isBinary = isbinaryfile.sync(content, fs.lstatSync(dest).size);

        if(!isBinary)
            content = content.toString(neoServ.encoding || config.encoding);
    }

    catch(e) {
        error(404);

        if(argument('d', 'debug'))
            console.log(e.stack);

        if(argument('b', 'benchmark'))
            console.log('Request sent in ' + (Date.now() - receivedRequest) + ' ms');

        return ;
    }

    if(typeof content !== 'undefined')
        if(send(request.treatUrl.substr(1), content))
            console.info(colors.blue('Delivered : ') + colors.green(request.url) + (request.url !== request.treatUrl ? ' [' + colors.cyan(request.treatUrl) + ']' : ''));

    if(argument('b', 'benchmark'))
        console.log('Request sent in ' + (Date.now() - receivedRequest) + ' ms');

}).listen(config.port);

console.log(colors.blue('Server started on port ' + colors.cyan(config.port)));
console.log('Press Ctrl+C to stop');
